File: ios/Sources/PamNative/Modules/BackgroundModule.swift
import Foundation
import UIKit

final class BackgroundModule: NativeModule, ClosableNativeModule {
    private var nextToken = 1
    private var tasks: [Int: UIBackgroundTaskIdentifier] = [:]

    func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        DispatchQueue.main.async {
            do {
                switch method {
                case "begin":
                    let token = self.nextToken
                    self.nextToken += 1
                    let identifier = UIApplication.shared.beginBackgroundTask(
                        withName: "PamNative-\(token)"
                    ) { [weak self] in
                        self?.end(token)
                    }
                    guard identifier != .invalid else {
                        throw BackgroundError("iOS denied the background task")
                    }
                    self.tasks[token] = identifier
                    completion(
                        .success,
                        (try? WireMap.encode(["token": .integer(Int64(token))])) ?? Data()
                    )
                case "end":
                    let values = try WireMap.decode(payload)
                    guard case let .integer(token)? = values["token"] else {
                        throw BackgroundError("Missing background token")
                    }
                    self.end(Int(token))
                    completion(.success, Data())
                default:
                    throw BackgroundError("Unknown background method \(method)")
                }
            } catch {
                completion(.failure, Data(error.localizedDescription.utf8))
            }
        }
    }

    private func end(_ token: Int) {
        guard let identifier = tasks.removeValue(forKey: token) else { return }
        UIApplication.shared.endBackgroundTask(identifier)
    }

    func close() {
        DispatchQueue.main.async {
            Array(self.tasks.keys).forEach(self.end)
        }
    }
}

private struct BackgroundError: LocalizedError {
    let message: String
    init(_ message: String) { self.message = message }
    var errorDescription: String? { message }
}

File: android/app/src/main/java/dev/pam/nativeapp/modules/BackgroundModule.kt
package dev.pam.nativeapp.modules

import android.content.Context
import android.os.PowerManager
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue
import java.util.concurrent.atomic.AtomicInteger

internal class BackgroundModule(private val context: Context) : NativeModule, AutoCloseable {
    private val nextToken = AtomicInteger(1)
    private val locks = mutableMapOf<Int, PowerManager.WakeLock>()

    override fun invoke(method: String, payload: ByteArray, completion: ModuleCompletion) {
        runCatching {
            when (method) {
                "begin" -> {
                    val values = WireMap.decode(payload)
                    val timeout = ((values["timeoutSeconds"] as? WireValue.Integer)?.value ?: 30L)
                        .coerceIn(1, 600)
                    val name = (values["name"] as? WireValue.Text)?.value.orEmpty().take(128)
                    val token = nextToken.getAndIncrement()
                    val lock = context.getSystemService(PowerManager::class.java)
                        .newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "PamNative:$name")
                    lock.acquire(timeout * 1_000)
                    synchronized(locks) { locks[token] = lock }
                    completion.complete(
                        ModuleResultStatus.SUCCESS,
                        WireMap.encode(mapOf("token" to WireValue.Integer(token.toLong()))),
                    )
                }
                "end" -> {
                    val token = (WireMap.decode(payload)["token"] as? WireValue.Integer)?.value?.toInt()
                        ?: error("Missing background token")
                    synchronized(locks) { locks.remove(token) }?.let {
                        if (it.isHeld) it.release()
                    }
                    completion.complete(ModuleResultStatus.SUCCESS, ByteArray(0))
                }
                else -> error("Unknown background method $method")
            }
        }.onFailure { error ->
            completion.complete(ModuleResultStatus.FAILURE, (error.message ?: "Background task failed").toByteArray())
        }
    }

    override fun close() {
        synchronized(locks) {
            locks.values.forEach { if (it.isHeld) it.release() }
            locks.clear()
        }
    }
}
